src/pstl_offload.cpp

```diff
@@ -96,59 +97,311 @@ __original_msize(void* __user_ptr)
+        static constexpr unsigned __ptr_shift = 11;
+        // current page size is at least 4K
+        assert(__get_memory_page_size() >= 1 << __ptr_shift);
+        return std::hash<void*>()((void*)((uintptr_t)__p >> __ptr_shift));
+    }
+};
+
+template<class T>
+struct __orig_free_allocator
+{
+    typedef T value_type;
+
+    __orig_free_allocator() = default;
+
+    template<class U>
+    constexpr __orig_free_allocator(const __orig_free_allocator <U>&) noexcept {}
+
+    T* allocate(std::size_t __n)
+    {
+        if (T *ptr = static_cast<T*>(std::malloc(__n * sizeof(T))))
+        {
+            return ptr;
+        }
+        throw std::bad_alloc();
+    }
+
+    void deallocate(T* __ptr, std::size_t) noexcept
+    {
+        __original_free(__ptr);
+    }
+};
+
+template<class T, class U>
+bool operator==(const __orig_free_allocator <T>&, const __orig_free_allocator <U>&) { return true; }
+template<class T, class U>
+bool operator!=(const __orig_free_allocator <T>&, const __orig_free_allocator <U>&) { return false; }
+
+static __spin_mutex __large_aligned_ptrs_map_mtx;
+
+class __large_aligned_ptrs_map
+{
+public:
+    struct __ptr_desc
+    {
+        __sycl_device_shared_ptr _M_device;
+        std::size_t _M_requested_number_of_bytes;
+    };
+
+private:
+    // Find sycl::device and requested size by user pointer. Use __orig_free_allocator to not
+    // call global delete during delete processing (overloaded global delete includes
+    // __unregister_ptr call).
+    using __map_ptr_to_object_prop =
+        std::unordered_map<void*, __ptr_desc, __hash_aligned_ptr, std::equal_to<void*>,
+        __orig_free_allocator<std::pair<void* const, __ptr_desc>>>;
+    __map_ptr_to_object_prop* _M_map;
+
+public:
+    // We suppose that all users of libpstloffload have dependence on it, so it's impossible to
+    // register >=4K-aligned USM memory before ctor of static objects in libpstloffload is executed.
+    // So, no need for special support for adding to not-yet-created __large_aligned_ptrs_map.
+    __large_aligned_ptrs_map()
+    {
+        std::scoped_lock __l(__large_aligned_ptrs_map_mtx);
+        _M_map = new __map_ptr_to_object_prop;
+    }
+
+    // Do not destroy (i.e., intentionally leak) _M_map to able use it after static object dtor is
+    // executed. Global free/delete/realloc/etc are overloaded, so we need to use it even after
+    // static object dtor has been executed.
+    ~__large_aligned_ptrs_map() { }
+
+    void
+    __register_ptr(void* __ptr, std::size_t __size, __sycl_device_shared_ptr __device_ptr)
+    {
+        assert(__is_ptr_page_aligned(__ptr));
+
+        std::scoped_lock __l(__large_aligned_ptrs_map_mtx);
+        [[maybe_unused]] auto __ret = _M_map->emplace(__ptr, __ptr_desc{std::move(__device_ptr), __size});
+        assert(__ret.second); // the pointer must be unique
+    }
+
+    // nullopt means "it's not our pointer"
+    static std::optional<__ptr_desc>
```

Comment:
Do we really need `optional`? C/C++ pointer is sort of optional itself. Is there any particular advantage to use optional?

Reply:
Discussed offline. Let's stick to optional right now